pkg/executables/cmk.go

```diff
@@ -192,7 +195,15 @@ func (c *Cmk) ValidateZonesPresent(ctx context.Context, zones []v1alpha1.CloudSt
 func (c *Cmk) ValidateDomainPresent(ctx context.Context, domain string) (v1alpha1.CloudStackResourceIdentifier, error) {
 	domainIdentifier := v1alpha1.CloudStackResourceIdentifier{Name: domain, Id: ""}
 	command := newCmkCommand("list domains")
-	applyCmkArgs(&command, withCloudStackName(domain), appendArgs("listall=true"))
+
+	if domain == rootDomain {
+		applyCmkArgs(&command, withCloudStackName(domain))
+	} else {
+		tokens := strings.Split(domain, domainDelimiter)
+		domainName := tokens[len(tokens)-1]
```

Comment:
This seems convoluted. If a user supplies a domain `foo/bar` can we not search `foo/bar` instead of `foo`? 

If `foo/bar` is an absolute reference to an existing domain, great, we have a match. 

In the event the user only supplies `bar` and we get back `ROOT/bar` and `ROOT/foo/bar` like the PR description suggests is possible then we error because the domain they've specified is ambiguous.

Reply:
The link below also suggests there are some other options that may alter the response structure such as `level`. Have we experimented with those as they may yield something thats more easily inspectable for ambiguous domains?

Reply:
> This seems a little convoluted. If a user supplies a domain `foo/bar` can we not search `foo/bar` instead of extracting and searching just `bar`?

Unfortunately, ACS 'list domains' API does not support querying by 'path'. That's why I have to parse the 'name' part to list all domains with it. Then I find a single domain that has the domain 'path'

> If `foo/bar` is an absolute reference to an existing domain, great, we have a match.
> 
> In the event the user only supplies `bar` and we get back `ROOT/bar` and `ROOT/foo/bar` like the PR description suggests is possible then we error because the domain they've specified is ambiguous.

Yes, in this case the pre-flight check returns an error that it discovered multiple domains based on the provided information.

Reply:
I see. Can someone search for a parent domain? 

E.g.

```
# Domains
# ROOT/foo/bar
# ROOT/foo/qux

$ cmk list domains foo listall=true # results in both ROOT/foo/bar and ROOT/foo/qux
```

Reply:
Spoke with Wonkun. 

We noted a few examples where modeling the code to make assumptions about what the user wants to do could be quite undesirable. 

Specifically, if the user has a complex path structure that exhibits the same domain name we're assuming we know which one they want which is wrong. If there's ambiguity the user needs to clear that up for us. 

Example
```
ROOT/foo
ROOT/qux/foo
ROOT/bar/foo
ROOT/aux/bar/foo
```
We would need to choose 1 of 4 options and we're essentially guessing what the user wants.

I proposed introducing a requirement to specify _either_ a path or a domain name using the same parameter and they must be distinguishable so handling code knows what to do. This could simply mean paths must be specified with a preceeding slash.

Using the example above, if a user supplies `/foo` its clear they want `ROOT/foo`. If they specify `foo` its a domain and its ambiguous so we would error. 

I feel being able to distinguish the users intent is imperative to modeling the correct behavior.

Reply:
> I see. Can someone search for any portion of a domain?
> 
> E.g.
> 
> ```
> # Domains
> # ROOT/foo/bar
> # ROOT/foo/qux
> 
> $ cmk list domains foo listall=true # results in both ROOT/foo/bar and ROOT/foo/qux
> ```

In this case, only "ROOT/foo" is returned.

Reply:
@chrisdoherty4 the CAPC team discussed this issue, and we believe that FQDN is the correct way to go. We can update the CAPC API description of the field to reflect that.

In this manner we won't have ambiguous logic regarding the resolution of Domain path.

`foo == ROOT/foo` at least for now

In the future we do mean to break out domain resolution into a Go type including Path and ID, and at that point we could even entertain the idea of a Name field which would fill the niche for usage of just `foo` as you've detailed above. Unfortunately, to change field names and or types, we'd need to introduce API version conversion webhooks.

These webhooks are on our roadmap, but not something we can implement in time for our customer's needs.

This PR will solve the use case of VM creation in a subdomain while not breaking backward compatibility.

Reply:
Having chatted offline we discussed sticking with an FQDN. In the case a single response is returned we still need to validate the complete path matches. This means users currently using the product _may_ need to correct to use an FQDN.

With respect to paths that I mistakenly thought we were trying to accommodate here (I think), its deferred as a separate piece of work. 

This behavior will create consistency in all situations and provide a pretty clear expectation for customers.